### PR TITLE
PERF: CategoricalIndex.get_loc should avoid expensive cast of .codes to int64

### LINKED ISSUE #20395: PERF: df.loc is 100x slower for CategoricalIndex than for normal Index
EDIT: After #21369 was merged the result of ``%timeit df2.loc['b']`` has improved to  **3.8 ms**.
EDIT: After #21618 was merged the result of ``%timeit df2.loc['b']`` has improved to  **3.3 ms**.
EDIT: After #21659 was merged the result of ``%timeit df2.loc['b']`` has improved to  **1.6 ms**.

#### Code Sample

```python
>>> n = 100_000
>>> df1 = pd.DataFrame(dict(A=range(n*3)), index=list('a'*n + 'b'*n + 'c'*n))
>>> df1.index.is_monotonic_increasing
True
>>> df2 = df1.copy()
>>> df2.index = pd.CategoricalIndex(df2.index)
>>> df2.index.is_monotonic_increasing
True
>>> %timeit df1.loc['b']
125 µs ± 2.95 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
>>> %timeit df2.loc['b']
13.8 ms ± 193 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
#### Problem description

Selecting on a ``CategoricalIndex`` is 100x slower than selecting on a normal ``Index``.

I've tested this on master ( a few days old)  and on v0.22, with same result for both versions. The speed is even worse than the speed for a full columns scan:

```python
>>> df3 = df2.reset_index()
>>> %timeit df3[df3['index'] == 'b']
6.58 ms ± 25.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

A guess is that the binary search is bypassed and a full index scan is being done + some extra stuff  so it's even slower than a normal full columns scan.

#### Expected Output

The output is as expected, but the speed is very slow for ``CategoricalIndex``.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: a7a7f8c1101aed1a9d37abbbcd80f77da414f0a8
python: 3.6.3.final.0
python-bits: 32
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0.dev0+870.ga7a7f8c
pytest: 3.3.1
pip: 9.0.1
setuptools: 38.2.5
Cython: 0.26.1
numpy: 1.13.3
scipy: 1.0.0
pyarrow: None
xarray: 0.10.0
IPython: 6.2.1
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.0
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: None
bs4: None
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #20395
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This is the the final puzzle for #20395.

The problem with the current implementation is that ``CategoricalIndex._engine`` constantly recodes int8 arrays to int64 arrays:

https://github.com/pandas-dev/pandas/blob/dc45fbafef172e357cb5decdeab22de67160f5b7/pandas/core/indexes/category.py#L365-L369

Notice the ``lambda: self.codes.astype('i8')`` part, which means that every time ``_engine.vgetter`` is called, an int64-array is created. This is very expensive and we would ideally want to just use the original array dtype (int8, int16 or whatever) always and avoid this conversion.

A complicating issue is that it is *not* enough to just avoid the int64 transformation, as ``array.searchsorted`` apparantly needs a dtype-compatible input or else it is *also* very slow:

```python
>>> n = 1_000_000
>>> ci = pd.CategoricalIndex(list('a' * n + 'b' * n + 'c' * n))
>>> %timeit ci.codes.searchsorted(1)  # search for left location of 'b'
7.38 ms   # slow
>>> code = np.int8(1)
>>> %timeit ci.codes.searchsorted(code)
2.57 µs  # fast
```

### Solution options

As CategoricalIndex.codes may be int8, int16, etc, the solution must be to (1) have an indexing engine for each integer dtype or an indexing engine that accepts all int types, not just int64 *and* (2) that the key must be translated into the same dtype as the codes array before calling searchsorted. So either:
1. Change Int64Engine to be a IntEngine (i.e. accept all integer dtypes)
2. Make new IntEngine classes, with the appropriate flexibility for accepting all integer dtypes, but defers to Int64 version if/when needed (e.g. if codes is int8, but we only have ``algos.is_monotonic_int64`` for checking monotonicity)
3. Do everything in Python.

I assume option 1 is not desired, and option 3 assumedly likewise. In the updated PR I've made a proposal in Cython, that attains the needed speed.

Benchmarks from ``asv_bench/indexing.py``

```
      before           after         ratio
     [dc45fbaf]       [bc03b8bd]
-     1.54±0.02ms         9.80±0μs     0.01  indexing.CategoricalIndexIndexing.time_get_loc_scalar('monotonic_incr')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

### CODE REVIEW

**Comment by WillAyd** [comment]
@topper-123 what issues were you running into with Cython? As far as generating the indexes goes I would think you could just add the appropriate types below:

https://github.com/pandas-dev/pandas/blob/2b136054db3b8f4c8d1aa33160c4bbe3b8ce940e/pandas/_libs/index_class_helper.pxi.in#L14

Not sure how you were planning to dynamically select the appropriate engine in `category.py` but as far as generating those engines in the Cython layer I think most of the foundation is already in place

**Comment by topper-123** [comment]
Ok, I think I understand the design now, at least. What is the templating language used and how is the pxi-file generated?

So would the correct approach be to:
* add int8_t, int16_t etc. to that list in index_helper_pxi.in,
* similarly add them to the list in algos.common_helper.pxi.in,
* ensure that CategoricalIndex._engine is of the correct type,
* avoid the  casting to int64 in  ``CategoricalIndex._engine``?

**Comment by topper-123** [comment]
Ok, got algos.common_helper.pxi.in compiling and works fine.

index_helper.pxi throws a compiler error. I added to the dtype list in index_helper.pxi.in:

```
          ('Int8', 'int8', 'int8_t'),
```

and get this error:

```cython
cythoning pandas\_libs/index.pyx to pandas\_libs\index.c

Error compiling Cython file:
------------------------------------------------------------
...

    cdef _maybe_get_bool_indexer(self, object val):
        cdef:
            ndarray[uint8_t, ndim=1, cast=True] indexer
            ndarray[intp_t, ndim=1] found
            ndarray[int8_t] values
                         ^
------------------------------------------------------------

pandas\_libs\index_class_helper.pxi:179:26: Invalid type.
````

It appears the compiler does not like the int8_t type. Using ``('UInt8', 'uint8', 'uint8_t'),`` instead compiles without errors.

Any ideas what's happening?

**Comment by jreback** [comment]
you need to do the imports of ``int8_t`` in ``index.pyx`` itself (see the top).

**Comment by topper-123** [comment]
Thanks, got everything compiling with Int8Engine, which is good.

However, when running, it complains there is no Int8HashTable. Seems like there isn't an easy way to create and Int8HashTable, is that correct?

```python
~\Documents\Python\pandasdev\pandasdev\pandas\_libs\index_class_helper.pxi in pandas._libs.index.Int8Engine._make_hash_table()
    212
    213     cdef _make_hash_table(self, n):
--> 214         return _hash.Int8HashTable(n)

AttributeError: module 'pandas._libs.hashtable' has no attribute 'Int8HashTable'
```

I tried looking into using a Int64HashTable in Int8Engine, but that seems to bring a different set of tradeoffs...

Is there an easy way to create a Int8HashTable, or is it reasonable to use a Int64HashTable in Int8Engine?

**Comment by jreback** [comment]
you might need to create the other hash tables  (definitions) as well
i think we only have 32 & 64 hits atm

**Comment by topper-123** [comment]
Ok, I've tried creating the int8 hash table, but got some issues. Everything compiles up till and including " add Int8Index", but that runs into the issue that no Int8HashTable existst. I've tried creating a Int8HashTable (see the commit "add Int8HashTable"), but I can't get  to compile.

It gives this compilation error:

```
Error compiling Cython file:
------------------------------------------------------------
...
        self.external_view_exists = True
        return self.ao

    cdef inline void append(self, int8_t x):

        if needs_resize(self.data):
                      ^
------------------------------------------------------------

pandas\_libs\hashtable_class_helper.pxi:287:23: no suitable method found

Error compiling Cython file:
------------------------------------------------------------
...


cdef class Int8HashTable(HashTable):

    def __cinit__(self, size_hint=1):
        self.table = kh_init_int8()
                                ^
------------------------------------------------------------

pandas\_libs\hashtable_class_helper.pxi:1291:33: undeclared name not builtin: kh_init_int8
```

I've tried messing with khash.pxd to get that kh_*_int8 working, but can't get it running.

What does khash.pxd do? It seems to in turn depend on khash_python.h, which only seems to have int64, int32 anf float64 version available. Must that be changed before getting a Int8HashTable working?

Guidance would be very much appreciated.

**Comment by jreback** [comment]
hmm, i guess khash doesn't support this either.

ok, so maybe let's do a precursor PR to do this properly

**Comment by topper-123** [comment]
Ok, just to confirm, so you/someone else will do the PR to get Int8HashTable etc. working, and I'll wait for that and afterwards I will implement Int8Engine etc?

**Comment by jreback** [comment]
@topper-123 well I suppose someone can do a PR to make other types of hashtables work. cc @pandas-dev/pandas-core

**Comment by topper-123** [comment]
That would be great. If someone makes the hash tables I will take this the rest of the way.

**Comment by topper-123** [comment]
I'v made a commit where I've solved the lack of Int8HashTable etc. by letting Int8Engine, Int16Engine and Int32Engine use Int64HashTable. All current tests pass and speed is good.

Is this an ok approach? I can then add some type specific tests if needed and update asv benchmarks.

**Comment by topper-123** [comment]
I've made a solution for the problem of not having Int8HashTable, Int16HashTable and Int32HashTable by using Int64HashTable in Int8Engine etc. all the tests pass performance is same as above.

Is this an ok approach? I could then add some tests and ASVs as needed.

A nice side effect BTW is that we now have and can use ``pd._libs.algos.is_monotonic_int8``, so we don't have to convert to int64 when checking for monoticity and uniqueness:

```python
>>> n = 1_000_000
>>> ci = pd.CategoricalIndex(list('a' * n + 'b' * n + 'c' * n))
# the below is called in ci.is_monotonic etc. in master
>>> %timeit pd._libs.algos.is_monotonic_int64(ci.codes.astype('int64'), False)
21.6 ms  # this PR
# the below is called in ci.is_monotonic etc. in this PR
>>> %timeit pd._libs.algos.is_monotonic_int8(ci.codes, False)
5.6 ms  # this PR
```

So all code that calls CategoricalIndex.is_monotonic/.is_unique will benefit also.

File: pandas/core/indexes/category.py [hunk 1]
```diff
@@ -429,10 +437,11 @@ def get_loc(self, key, method=None):
         >>> non_monotonic_index.get_loc('b')
         array([False,  True, False,  True], dtype=bool)
         """
-        codes = self.categories.get_loc(key)
-        if (codes == -1):
-            raise KeyError(key)
```
Comment (topper-123) [inline pandas/core/indexes/category.py]: codes can never be -1. If key is not in self.categories, a KeyError was always raised.

File: pandas/core/indexes/category.py [hunk 2]
```diff
@@ -366,7 +370,7 @@ def argsort(self, *args, **kwargs):
     def _engine(self):
 
         # we are going to look things up with the codes themselves
-        return self._engine_type(lambda: self.codes.astype('i8'), len(self))
```
Comment (jreback) [inline pandas/core/indexes/category.py]: I don't think you need any changes in cython if you simply so: ``.astype('i8', copy=False)``, which is the cause of the perf issue, no?

its still not as nice as actually using type specific hashtables though (which this PR is not addressing)

File: pandas/core/indexes/category.py [hunk 3]
```diff
@@ -429,10 +437,11 @@ def get_loc(self, key, method=None):
         >>> non_monotonic_index.get_loc('b')
         array([False,  True, False,  True], dtype=bool)
         """
-        codes = self.categories.get_loc(key)
-        if (codes == -1):
-            raise KeyError(key)
-        return self._engine.get_loc(codes)
+        code = self.categories.get_loc(key)
+        # dtype must be same as dtype for codes for searchsorted not lose speed
```
Comment (jreback) [inline pandas/core/indexes/category.py]: blank line between

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/category.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/category.py]: then you have something else going on

Comment (topper-123) [reply pandas/core/indexes/category.py]: The numpy docs say about array.astype:

    By default, astype always returns a newly allocated array. If this is set
    to false, and the dtype, order, and subok requirements are satisfied,
    the input array is returned instead of a copy.

So, calling astype(..., copy=False) will only avoid returning a copy when the dtype of codes is int64, i.e. in practice never for Categoricals.

**Comment by topper-123** [comment]
Hey, ``array.astype('int64', copy=False)`` won't work as explained above.

I don't see any good way forward other than someone commiting to make an Int8/16/32HashTable at some point. The question is thought if this PR needs to wait for that, or can be pulled in now and the relevant lines be changed when the hash table is implemented. 

I argue pulling this in now, as index operations are critical for performance and that would allow me to pursue a few more index-related performance issues.

File: pandas/core/indexes/category.py (hunk 2 above)
Comment (jreback) [reply pandas/core/indexes/category.py]: so try using ``algos.ensure_int64(values)`` I find this extremely odd that you have to do the if *inside* cython.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by topper-123** [comment]
``algos.ensure_int64(values)`` ends up calling ``values.astype(np.int64)``, and therefore also causes the slowdown for int8/16/32 arrays that we want to avoid.

A *technically* possible solution to avoid that if statement while keeping the speed improvements would be to input a int64 array into _engine:

```python
    def _engine(self):
        codes = self.codes.astype('int64')
        return self._engine_type(lambda: codes, len(self))
```

But that would keep an int64 array in memory, in addition to an int8 array, so doesn't make any sense memorywise. We want to just use the existing int8 array.

I made another solution, where I've moved this specific implementation to index_class_helper.pxi.in. This makes its location more specific to the problem it solves.

File: pandas/_libs/index_class_helper.pxi.in [hunk 4]
```diff
@@ -36,10 +40,17 @@ cdef class {{name}}Engine(IndexEngine):
...
         {{if name == 'Object'}}
         return _hash.PyObjectHashTable(n)
+        {{elif name in {'Int8', 'Int16', 'Int32'} }}
+        return _hash.Int64HashTable(n)
         {{else}}
         return _hash.{{name}}HashTable(n)
         {{endif}}
 
+    {{if name in {'Int8', 'Int16', 'Int32'} }}
+    cpdef _call_map_locations(self, values):
```
Comment (jreback) [inline pandas/_libs/index_class_helper.pxi.in]: ok can you add a bit comment here on what is going on

File: pandas/core/indexes/category.py [hunk 5]
```diff
@@ -72,7 +72,11 @@ class CategoricalIndex(Index, accessor.PandasDelegate):
     """
 
     _typ = 'categoricalindex'
-    _engine_type = libindex.Int64Engine
+
+    @property
+    def _engine_type(self):
+        type_name = self.codes.dtype.name.capitalize()
```
Comment (jreback) [inline pandas/core/indexes/category.py]: can you add a comment here

File: pandas/_libs/index_class_helper.pxi.in [hunk 6]
```diff
@@ -12,9 +12,13 @@ WARNING: DO NOT edit .pxi FILE directly, .pxi is generated from .pxi.in
 
 # name, dtype, ctype
 dtypes = [('Float64', 'float64', 'float64_t'),
-          ('UInt64', 'uint64', 'uint64_t'),
           ('Int64', 'int64', 'int64_t'),
-          ('Object', 'object', 'object')]
+          ('Int32', 'int32', 'int32_t'),
```
Comment (jreback) [inline pandas/_libs/index_class_helper.pxi.in]: add for other UInt types (8,16,32) as well

**Review by jreback — CHANGES_REQUESTED** [review]
ok looks better, with enough comments this would be ok. pls add a whatsnewnote, and a battery of tests which exercise all dtypes on this (ideally add asv which exercise all this as well)

**Comment by topper-123** [comment]
I've addressed the comments. Haven't made any new ASVs though, will look at it tomorrow, but  many of the new index engines are not currently used, so have to think about how to speed test them.

File: pandas/_libs/index_class_helper.pxi.in [hunk 7]
```diff
@@ -30,6 +30,8 @@ dtypes = [('Float64', 'float64', 'float64_t'),
 
 cdef class {{name}}Engine(IndexEngine):
 
+    _dtype = '{{dtype}}'
```
Comment (topper-123) [inline pandas/_libs/index_class_helper.pxi.in]: For testing purposes (see test_engine.py), we need the dtype that is compatible with this engine + it's useful information to have in general.

**Comment by topper-123** [comment]
The travis failure is the unexpected ResourceWarning, so unrelated to this PR.

**Comment by topper-123** [comment]
I added a commit with ASVs. I check ``*Engine.get_loc`` and get these result:

```
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For pandas commit hash 9df4167d:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 50.00%] ··· Running indexing_engines.EngineIndexing.time_get_loc                                                      4.07±0.4μs;...
[ 50.00%] · For pandas commit hash 03707400:
[ 50.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[100.00%] ··· Running indexing_engines.EngineIndexing.time_get_loc                                                              failed
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```
The failure at commit hash 03707400 is ok, as most indexing engines were not made yet there (Int8Engine etc. are new). The speed of 4.07μs is as expected.

I didn't make timing checks for ``is_monotonic*`` and ``is_unique``, as those are both cached, so it won't give any benefits timing them.

Is there anything else that should be timed?

**Comment by TomAugspurger** [comment]
Sorry I havent followed this closely, but you’ve added new indexing engines? Can you measure the size of the unzipped wheel before and after your changes?

________________________________

**Comment by topper-123** [comment]
Hey @TomAugspurger. Yes, CategoricalIndex typically work on int8 or int16 datatypes, so we need a corresponding indexing engine (Int8Engine if codes dtype is int8, etc.) to avoid converting to int64 and slowing down.

Strictly speaking, this PR only requires we do integer engines, but It seems sensible to me to do engines for the smaller UInts and Floats also.

I haven't worked with wheels myself, how do I find/create them? The size of the build folder was 78,5 MB before and 80,5 MB after this PR.

**Comment by TomAugspurger** [comment]
python setup.py bdist_wheel will create a wheel. That can be unzipped to
get the distribution size.

I just want to raise the cost of templating across many dtypes, since
that's what tends to blow up the distribution size.

**Comment by topper-123** [comment]
Ok, I get unzipped 24.5 MB (master) vs 24.8 MB (this PR)

**Comment by topper-123** [comment]
Is this ok?

**Comment by topper-123** [comment]
hey @jreback , is this ok to pull in?

File: pandas/_libs/index_class_helper.pxi.in [hunk 8]
```diff
@@ -12,16 +12,26 @@ WARNING: DO NOT edit .pxi FILE directly, .pxi is generated from .pxi.in
...
+          ('Float32', 'float32', 'float32_t'),
           ('Int64', 'int64', 'int64_t'),
-          ('Object', 'object', 'object')]
+          ('Int32', 'int32', 'int32_t'),
+          ('Int16', 'int16', 'int16_t'),
+          ('Int8', 'int8', 'int8_t'),
+          ('UInt64', 'uint64', 'uint64_t'),
+          ('UInt32', 'uint32', 'uint32_t'),
+          ('UInt16', 'uint16', 'uint16_t'),
+          ('UInt8', 'uint8', 'uint8_t'),
```
Comment (jorisvandenbossche) [inline pandas/_libs/index_class_helper.pxi.in]: Are the lower Uint engines actually used somewhere in the code? (I think you only changed things in Categorical, and those only use the Int.. engines?)

So if they are not used, should we actually add them?

File: pandas/_libs/index_class_helper.pxi.in [hunk 9]
```diff
@@ -36,10 +46,33 @@ cdef class {{name}}Engine(IndexEngine):
     cdef _make_hash_table(self, n):
         {{if name == 'Object'}}
         return _hash.PyObjectHashTable(n)
+        {{elif name in {'Int8', 'Int16', 'Int32'} }}
+        # {{name}}HashTable is not available, so we use Int64HashTable
+        return _hash.Int64HashTable(n)
```
Comment (jorisvandenbossche) [inline pandas/_libs/index_class_helper.pxi.in]: Since all the Int8Engine, Int16Engine, .. are actually still using the same Int64HashTable, is it actually needed to create those different engines? Can the `Int64Engine` (or IntEngine) not simply always do the `ensure_int64` in `_call_map_locations`? (which seems to be the only difference between Int8Engine and Int64Engine, from a quick skim, didn't look in detail)

**Review by jorisvandenbossche — COMMENTED** [review]
Few comments. I didn't fully follow the discussion that lead to the current state of the PR, sorry if I missed some things (it might be good to give an updated summary of the rationale of the PR and its implementation, in case the top post is no longer up to date)

File: pandas/_libs/index_class_helper.pxi.in (hunk 8 above)
Comment (topper-123) [reply pandas/_libs/index_class_helper.pxi.in]: The lower UInt engine are not used in the code ATM. I added them by suggestion from @jreback. I don't have a super strong preference on whether to add them now, but I can see benefits to aving them, e.g.:
* if someone want to make specialized indexes based on UInts they'd be nice to have.
* We don't only add UInt8/16/32 engines but also various UInt8/16/32 algorithms, that could be beneficial to have around for when the need is there.

all in all I'm weakly positive on adding them now, but will let you core devs decide on it.

File: pandas/_libs/index_class_helper.pxi.in (hunk 9 above)
Comment (topper-123) [reply pandas/_libs/index_class_helper.pxi.in]: The requirement needed to achieve the speedup is avoiding type conversion. I'm pretty sure that ``ensure_int64`` will convert a int8 array, so lose speeed? 

I'm not at the coding computer ATM, but I can double check how ``ensure_int64`` works tonight.

Comment (jorisvandenbossche) [reply pandas/_libs/index_class_helper.pxi.in]: > I'm pretty sure that ensure_int64 will convert a int8 array to int64, so lose speeed?

But the Int8Engine, Int16Engine, .. are actually calling `ensure_int64` in the current PR ? (not `ensure_int8`, ...)

Comment (topper-123) [reply pandas/_libs/index_class_helper.pxi.in]: the ``self.mapping.map_locations(algos.ensure_int64(values))`` is only converting *``values``*, i.e. typically a scalar, so it's very fast.

It would be a different an dslower matter if we had converted the array, but we don't do that...

Comment (jorisvandenbossche) [reply pandas/_libs/index_class_helper.pxi.in]: Ah, OK, I see in `_get_index_values` it is `ensure_{{dtype}}` that is used.

File: asv_bench/benchmarks/indexing_engines.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        values = list([1] * N + [2] * N + [3] * N)
+        array_ = {
+            'monotonic_incr': np.array(values, dtype=engine._dtype),
+            'monotonic_decr': np.array(list(reversed(values)),
+                                       dtype=engine._dtype),
+            'non_monotonic': np.array([1, 2, 3] * N, dtype=engine._dtype),
+        }[index_type]
+
+        self.data = engine(lambda: array_, len(array_))
+        self.int_scalar = 2
```
[the remark this answers is not in the document]
Comment (topper-123) [reply asv_bench/benchmarks/indexing_engines.py]: Ok, done.

File: doc/source/whatsnew/v0.24.0.txt [hunk 11]
```diff
@@ -497,15 +497,17 @@ Removal of prior version deprecations/changes
...
-- Improved performance of :func:`DataFrame.set_index` with columns consisting of :class:`Period` objects (:issue:`21582`,:issue:`21606`)
+- Slicing Series and Dataframe with an monotonically increasing :class:`CategoricalIndex`
+  is now very fast and has speed comparable to slicing with Int64Index. Memory
+  consumption from such slicing operations is likewise much lower.
+  The speed increase is both when indexing by label (using .loc) and position(.iloc) (:issue:`20395`)
+- Slicing a ``CategoricalIndex`` itself (i.e. ``ci[1000:2000]``) shows similar speed improvements as above (:issue:`21659`)
 - Improved performance of membership checks in :class:`Categorical` and :class:`CategoricalIndex`
   (i.e. ``x in cat``-style checks are much faster). :meth:`CategoricalIndex.contains`
   is likewise much faster (:issue:`21369`, :issue:`21508`)
+- Improved performance of :func:`Series.describe` in case of numeric dtypes (:issue:`21274`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: generally pls try to avoid changing orthogonal things.

File: pandas/tests/indexes/conftest.py [hunk 12]
```diff
@@ -45,3 +46,15 @@ def zero(request):
...
     return request.param
+
+
+@pytest.fixture(
+    params=[
+        'Int64', 'Int32', 'Int16', 'Int8',
+        'UInt64', 'UInt32', 'UInt16', 'UInt8',
+        'Float64', 'Float32',
+    ])
+def num_engine(request):
```
Comment (jreback) [inline pandas/tests/indexes/conftest.py]: make this more descriptive, numeric_indexing_engine

File: pandas/tests/indexes/test_category.py [hunk 13]
```diff
@@ -1111,3 +1112,27 @@ def test_take_invalid_kwargs(self):
         msg = "the 'mode' parameter is not supported"
         tm.assert_raises_regex(ValueError, msg, idx.take,
                                indices, mode='clip')
+
+
+class TestCategoricalIndexEngine(object):
+
+    @pytest.mark.parametrize('nbits', [8, 16, 32, 64])
+    def test_engine_type(self, nbits):
```
Comment (jreback) [inline pandas/tests/indexes/test_category.py]: woa, this is very complicated. why is this necessary? can you make much more clear

File: pandas/tests/indexes/test_engine.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+import pytest
+
+import pandas as pd
+from pandas._libs.index import (Int64Engine, UInt64Engine,
+                                Float64Engine, ObjectEngine)
+
+
+class TestNumericEngine(object):
+
+    @pytest.mark.parametrize('data', [[0, 1, 2]])
```
Comment (jreback) [inline pandas/tests/indexes/test_engine.py]: why is this parameterized over data?

File: pandas/tests/indexes/test_engine.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        elif issubclass(index.dtype.type, np.unsignedinteger):
+            assert isinstance(index._engine, UInt64Engine)
+        elif issubclass(index.dtype.type, np.floating):
+            assert isinstance(index._engine, Float64Engine)
+        else:
+            raise TypeError("unexpected dtype {}".format(index.dtype))
+
+    @pytest.mark.parametrize('data', [[0, 1, 2]])
+    def test_is_monotonic_ordered(self, data, num_engine):
+        codes = np.array(data, dtype=num_engine._dtype)
```
Comment (jreback) [inline pandas/tests/indexes/test_engine.py]: why do you have multiple tests for monotonic? and not monotonic?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.24.0.txt (hunk 11 above)
Comment (topper-123) [reply doc/source/whatsnew/v0.24.0.txt]: I can see how that may be confusing. I'll limit doing this from now on, but this was just a spelling error, that I corrected, as it was adjacent to my text..

File: pandas/tests/indexes/test_category.py (hunk 13 above)
Comment (topper-123) [reply pandas/tests/indexes/test_category.py]: Yeah, I've made it simpler.

File: pandas/tests/indexes/test_engine.py (hunk 14 above)
Comment (topper-123) [reply pandas/tests/indexes/test_engine.py]: Fixed. You start out with a certain structure and forget to check if is sensible for the end product...

File: pandas/tests/indexes/test_engine.py (hunk 15 above)
Comment (topper-123) [reply pandas/tests/indexes/test_engine.py]: Yeah, there just one data to check for. Fixed.

**Comment by topper-123** [comment]
Changes according to comments.

**Comment by topper-123** [comment]
Is this ok to merge in?

**Comment by gfyoung** [comment]
ping @jreback

**Comment by jreback** [comment]
will have a look

**Comment by jreback** [comment]
not sure of the state of thIS PR, needs a rebase

**Comment by gfyoung** [comment]
@jreback : I believe this PR is ready for review see [above](https://github.com/pandas-dev/pandas/pull/21699#issuecomment-415957938).  However, time as a resource is limited, and this one got overlooked unfortunately.

@topper-123 : Sorry for the delay, but can you rebase this?

**Comment by topper-123** [comment]
Rebased and green.

**Comment by gfyoung** [comment]
@topper-123 : Thanks!  Can you double check if you addressed all comments?  IINM, I see a few that might need attention...

Feel free to resolve those that you think have completed discussion.
